src/main/lib/osu-file-parser/OsuParser.ts

```diff
@@ -1,125 +1,345 @@
+import { access } from "../fs-promises";
+import { fail, ok } from "../rust-like-utils-backend/Result";
+import { assertNever } from "../tungsten/assertNever";
+import { OsuFile } from "./OsuFile";
 
 const bgFileNameRegex = /.*"(?<!Video.*)(.*)".*/;
 const beatmapSetIDRegex = /([0-9]+) .*/;
 
-
-
-type FileState = 'Initial'
-  | 'NextState'
-  | 'General'
-  | 'Editor'
-  | 'Metadata'
-  | 'Difficulty'
-  | 'Events'
-  | 'TimingPoints'
-  | 'Colours'
-  | 'HitObjects';
+type FileState =
+  | "Initial"
+  | "NextState"
+  | "General"
+  | "Editor"
+  | "Metadata"
+  | "Difficulty"
+  | "Events"
+  | "TimingPoints"
+  | "Colours"
+  | "HitObjects";
 
 const OFFSET = 0;
 const BPM = 1;
 
-
-
 type Table<T> = Map<ResourceID, T>;
 
-export type DirParseResult = Promise<Result<[Table<Song>, Table<AudioSource>, Table<ImageSource>], string>>
+export type DirParseResult = Promise<
+  Result<[Table<Song>, Table<AudioSource>, Table<ImageSource>], string>
+>;
+
+// Overriding Buffer prototype because I'm lazy.
+// Should probably get moved to another file, or make a wrapper instead.
+
+Buffer.prototype.read_bytes = function (n) {
+  const out = this.slice(this.pos, this.pos + n);
+  this.pos += n;
+  return out;
+};
+
+Buffer.prototype.read_string = function () {
+  const empty_check = this.read_u8();
+  if (empty_check == 0) return "";
+
+  const len = this.read_uleb128();
+  return this.read_bytes(len).toString("utf-8");
+};
+
+Buffer.prototype.read_hash = function () {
+  const empty_check = this.read_u8();
+  if (empty_check == 0) return "00000000000000000000000000000000";
+
+  let len = this.read_uleb128();
+  if (len > 32) len = 32;
+
+  return this.read_bytes(len).toString("utf-8");
+};
+
+Buffer.prototype.read_uleb128 = function () {
+  let result = 0;
+  let shift = 0;
+  let byte = 0;
+
+  do {
+    byte = this.read_u8();
+    result |= (byte & 0x7f) << shift;
+    shift += 7;
+  } while (byte & 0x80);
+
+  return result;
+};
+
+Buffer.prototype.read_f64 = function () {
+  const bytes = this.read_bytes(8);
+  return bytes.readDoubleLE(0);
+};
+
+Buffer.prototype.read_f32 = function () {
+  const bytes = this.read_bytes(4);
+  return bytes.readFloatLE(0);
+};
+
+Buffer.prototype.read_u64 = function () {
+  const low = this.read_u32();
+  const high = this.read_u32();
+  return (high << 32) | low;
+};
+
+Buffer.prototype.read_u32 = function () {
+  const bytes = this.read_bytes(4);
+  return bytes.readUInt32LE(0);
+};
+
+Buffer.prototype.read_u16 = function () {
+  const bytes = this.read_bytes(2);
+  return bytes.readUInt16LE(0);
+};
+
+Buffer.prototype.read_u8 = function () {
+  return this.read_bytes(1)[0] & 0xff;
+};
 
 export class OsuParser {
-  static async parseDir(dir: string, update?: (i: number, total: number, file: string) => any): DirParseResult {
-    if (!await access(dir, fs.constants.R_OK)) {
-      return fail('Directory does not exists.');
+  static async parseDb(
```

Comment:
Please change `parseDb` to fully qualified name `parseDatabase` because the whole code base uses fully qualified names. `dir` being exception...

Reply:
replaced